llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp

```diff
@@ -634,6 +634,61 @@ void AMDGPUDAGToDAGISel::Select(SDNode *N) {
   case ISD::BUILD_VECTOR: {
     EVT VT = N->getValueType(0);
     unsigned NumVectorElts = VT.getVectorNumElements();
+
+    auto IsSplatAllZeros = [this](SDNode *N) -> bool {
+      if (ISD::isConstantSplatVectorAllZeros(N))
+        return true;
+
+      // Types may have legalized by stripping the 16 bit multi-element vector
```

Comment:
It would be better to adjust the legalizer to do this in the first place

Reply:
As in, allow vector types as long as they splat 0?

Reply:
Even if they don't splat 0

Reply:
The splatness isn't significant, and if we're going to have 64-bit vector elements we should have made them legal throughout the DAG, not just during selection